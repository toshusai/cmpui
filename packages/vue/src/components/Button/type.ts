export const ButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type ButtonVariant = (typeof ButtonVariant)[keyof typeof ButtonVariant];

export const ButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type ButtonSize = (typeof ButtonSize)[keyof typeof ButtonSize];

export type ButtonProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
};
