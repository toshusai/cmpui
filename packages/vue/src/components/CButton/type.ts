export const CButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type CButtonVariant =
  (typeof CButtonVariant)[keyof typeof CButtonVariant];

export const CButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type CButtonSize = (typeof CButtonSize)[keyof typeof CButtonSize];

export type CButtonProps = {
  variant?: CButtonVariant;
  size?: CButtonSize;
  disabled?: boolean;
};
