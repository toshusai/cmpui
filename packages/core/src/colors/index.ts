export * from "./blendColors";
export * from "./chooseHigherContrast";
export * from "./componentToHex";
export * from "./const";
export * from "./getContrast";
export * from "./getHighContrast";
export * from "./getLuminance";

export * from "./hexToHsv";

export * from "./hexToRgba";
export * from "./hsvToHex";
export * from "./hsvToRgb";
export * from "./hsvaToRgba";
export * from "./isValidColor";
export * from "./isValidHex";
export * from "./isValidHex6";
export * from "./rgbToCss";
export * from "./rgbToHex";

export * from "./rgbToHsv";
export * from "./rgbaArrayToRgba";
export * from "./rgbaToCss";
export * from "./srgbToLinear";
export * from "./types";

export * from "./useHighContrastColor"