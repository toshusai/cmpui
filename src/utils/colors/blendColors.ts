import { RGB } from "./types/RGB";

export function blendColors(
  foregroundColor: RGB,
  backgroundColor: RGB,
  alpha: number,
): RGB {
  return {
    r: Math.round(foregroundColor.r * alpha + backgroundColor.r * (1 - alpha)),
    g: Math.round(foregroundColor.g * alpha + backgroundColor.g * (1 - alpha)),
    b: Math.round(foregroundColor.b * alpha + backgroundColor.b * (1 - alpha)),
  };
}
