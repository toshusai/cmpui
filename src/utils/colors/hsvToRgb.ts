import { HSV, RGB } from "./types";

/**
 * Reference:
 * - Algorithm from: https://en.wikipedia.org/wiki/HSL_and_HSV#HSV_to_RGB
 */
export function hsvToRgb(hsv: HSV): RGB {
  let r = 0;
  let g = 0;
  let b = 0;

  const h = hsv.h;
  const s = hsv.s;
  const v = hsv.v;

  const i = Math.floor(h / 60);
  const f = h / 60 - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));

  switch (i % 6) {
    case 0:
      r = v;
      g = t;
      b = p;
      break;
    case 1:
      r = q;
      g = v;
      b = p;
      break;
    case 2:
      r = p;
      g = v;
      b = t;
      break;
    case 3:
      r = p;
      g = q;
      b = v;
      break;
    case 4:
      r = t;
      g = p;
      b = v;
      break;
    case 5:
      r = v;
      g = p;
      b = q;
      break;
  }

  return {
    r: Math.round(r * 255),
    g: Math.round(g * 255),
    b: Math.round(b * 255),
  };
}
