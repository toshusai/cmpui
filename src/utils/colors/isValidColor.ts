export function isValidColor(color: string) {
  const s = new Option().style;
  s.color = color;
  return s.color !== "";
}
