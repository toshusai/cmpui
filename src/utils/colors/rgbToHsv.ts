import { HSV, RGB } from "./types";

/**
 *
 * References:
 * - https://gist.github.com/mjackson/5311256
 * - https://en.wikipedia.org/wiki/HSL_and_HSV
 * @param r
 * @param g
 * @param b
 * @returns
 */
export function rgbToHsv(rgb: RGB): HSV {
  const red = rgb.r / 255;
  const green = rgb.g / 255;
  const blue = rgb.b / 255;

  const max = Math.max(red, green, blue);
  const min = Math.min(red, green, blue);
  const value = max;

  const chroma = max - min;
  const saturation = value === 0 ? 0 : chroma / value;

  let hue = 0;
  if (chroma === 0) {
    hue = 0;
  } else {
    switch (value) {
      case red:
        hue = (green - blue) / chroma + (green < blue ? 6 : 0);
        break;
      case green:
        hue = (blue - red) / chroma + 2;
        break;
      case blue:
        hue = (red - green) / chroma + 4;
        break;
    }
    hue /= 6;
  }

  return { h: hue * 360, s: saturation, v: value };
}
