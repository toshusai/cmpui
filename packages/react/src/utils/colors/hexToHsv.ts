import { hexToRgba } from "./hexToRgba";
import { rgbToHsv } from "./rgbToHsv";
import { HSVA } from "./types/HSVA";

export function hexToHsv(hex: string): HSVA {
  const rgb = hexToRgba(hex);
  return { ...rgbToHsv(rgb), a: rgb.a };
}
