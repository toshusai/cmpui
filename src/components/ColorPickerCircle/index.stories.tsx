import type { Meta, StoryObj } from "@storybook/react";
import { useCallback, useState } from "react";
import { ColorPickerCircle } from ".";
import { HSVA } from "../..";
import { hexToHsv } from "../../utils/colors/hexToHsv";

const meta: Meta<typeof ColorPickerCircle> = {
  title: "Components/Colors/ColorPickerCircle",
  component: ColorPickerCircle,
};

export default meta;
type Story = StoryObj<typeof ColorPickerCircle>;

export const Default: Story = {
  render: function Render() {
    const hsv = hexToHsv("#c649ba");
    const [saturation, setSaturation] = useState(hsv.s);
    const [value, setValue] = useState(hsv.v);
    const [hue, setHue] = useState(hsv.h);
    const [alpha, setAlpha] = useState(hsv.a);

    const handleChange = useCallback((hsva: HSVA) => {
      setHue(hsva.h);
      setSaturation(hsva.s);
      setValue(hsva.v);
      setAlpha(hsva.a);
    }, []);

    return (
      <ColorPickerCircle
        hue={hue}
        onChange={handleChange}
        alpha={alpha}
        saturation={saturation}
        value={value}
      />
    );
  },
};
