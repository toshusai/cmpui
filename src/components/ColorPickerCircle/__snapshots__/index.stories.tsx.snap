// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors/ColorPickerCircle Default smoke-test 1`] = `
<div class="cmpui_color-picker-circle__root">
  <div>
    <div class="cmpui_hue-circle__root"
         tabindex="0"
         style="--cmpui-width: 24px; --cmpui-size: 192px;"
    >
      <div class="cmpui_hue-circle__hue"
           style="width: 192px; height: 192px;"
      >
      </div>
      <div x="27.8363521006857"
           y="46.911130537989536"
           class="cmpui_circle__root cmpui_hue-circle__nob"
           style="--cmpui-circle-nob-fill: rgb(255,0,231); --cmpui-circle-nob-radius: 8px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 19.8364px; top: 38.9111px;"
      >
      </div>
      <div class="cmpui_hue-circle__inner">
      </div>
    </div>
    <div class="cmpui_color-picker-circle__sv-picker-parent">
      <div class="cmpui_sv-picker__root"
           tabindex="0"
      >
        <canvas width="93.33809511662427"
                height="93.33809511662427"
                style="width: 93.3381px; height: 93.3381px;"
        >
        </canvas>
        <div class="cmpui_circle__root cmpui_sv-picker__nob"
             x="58.92556509887897"
             y="20.86380949665719"
             style="--cmpui-circle-nob-fill: rgb(198,73,186); --cmpui-circle-nob-radius: 8px; --cmpui-circle-nob-stroke-color: rgb(255,255,255); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(255, 255, 255); left: 50.9256px; top: 12.8638px;"
        >
        </div>
      </div>
    </div>
  </div>
  <div class="cmpui_color-slider__root"
       tabindex="0"
  >
    <div style="position: relative; overflow: hidden; cursor: pointer; width: 192px; height: 20px;">
      <div class="alpha-checker-board"
           style="width: 192px; height: 20px;"
      >
        <div style="width: 100%; height: 100%; background: rgb(198, 73, 186); mask-image: linear-gradient(to right, transparent 0%, black 100%);">
        </div>
      </div>
      <div x="192"
           y="10"
           class="cmpui_circle__root alpha-checker-board"
           style="--cmpui-circle-nob-fill: white; --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 185.333px; top: 3.33333px; --cmpui-block-size: 4px; pointer-events: none;"
      >
      </div>
      <div x="192"
           y="10"
           class="cmpui_circle__root cmpui_color-slider__nob"
           style="--cmpui-circle-nob-fill: rgba(198,73,186,1); --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 185.333px; top: 3.33333px;"
      >
      </div>
    </div>
  </div>
</div>
`;
